optuna/search_space/intersection.py

```diff
@@ -1,13 +1,55 @@
+from __future__ import annotations
+
 from collections import OrderedDict
 import copy
 from typing import Dict
-from typing import Optional
+from typing import List
+from typing import Tuple
 
 import optuna
 from optuna.distributions import BaseDistribution
 from optuna.study import Study
 
 
+def _calculate(
+    trials: List[optuna.trial.FrozenTrial],
+    include_pruned: bool = False,
+    search_space: Dict[str, BaseDistribution] | None = None,
+    cursor: int = -1,
+) -> Tuple[Dict[str, BaseDistribution] | None, int]:
+    states_of_interest = [
+        optuna.trial.TrialState.COMPLETE,
+        optuna.trial.TrialState.WAITING,
+        optuna.trial.TrialState.RUNNING,
+    ]
+
+    if include_pruned:
+        states_of_interest.append(optuna.trial.TrialState.PRUNED)
+
+    trials_of_interest = [trial for trial in trials if trial.state in states_of_interest]
+
+    next_cursor = trials[-1].number + 1 if len(trials) > 0 else -1
```

Comment:
The original code checks the length of filtered trials.

```suggestion
    next_cursor = trials[-1].number + 1 if len(trials_of_interest) > 0 else -1
```

Reply:
Should we also get the number from this the last filtered trials? Maybe we should fix tests.

Reply:
Ideally, I should add some tests, but it seems like it would be quite complicated, so can I leave this problem open in this PR?

Reply:
You're right. Actually, current tests seem sufficient, and getting the number from `trials` seems fine in terms of correctness.